File: server/lib/jwt-secure.ts
/**
 * RemitFlow — Secure JWT Utilities
 * ═══════════════════════════════════════════════════════════════════════════
 * Replaces direct `jose` usage with hardened wrappers that:
 *   1. Reject JWE tokens with compression (CVE-2024-28176 mitigation)
 *   2. Enforce strict algorithm allowlists (no "alg: none" attacks)
 *   3. Enforce short token expiry windows
 *   4. Add structured logging for all auth failures
 *
 * CVE-2024-28176: jose JWE Decompression DoS — fixed in jose >=4.15.5
 *   Mitigation: We use jose 6.1.0 (patched) AND explicitly reject compressed JWEs.
 */

import { jwtVerify, SignJWT, decodeJwt, type JWTPayload } from "jose";

// ─── Constants ────────────────────────────────────────────────────────────────

/** Strict algorithm allowlist — never allow "none" or weak algorithms */
const ALLOWED_ALGORITHMS = ["RS256", "RS384", "RS512", "ES256", "ES384", "ES512"] as const;

/** Maximum token lifetime in seconds (1 hour) */
const MAX_TOKEN_AGE_SECONDS = 3600;

/** Maximum JWE payload size to prevent decompression bombs (CVE-2024-28176) */
const MAX_JWE_PAYLOAD_BYTES = 65_536; // 64 KiB

// ─── Types ────────────────────────────────────────────────────────────────────

export interface RemitFlowJWTPayload extends JWTPayload {
  sub: string;
  userId?: number;
  roles?: string[];
  sessionId?: string;
  keycloakId?: string;
}

export interface JWTVerifyResult {
  payload: RemitFlowJWTPayload;
  isValid: boolean;
  error?: string;
}

// ─── Secure Verify ────────────────────────────────────────────────────────────

/**
 * Verify a JWT with strict security controls.
 * - Rejects compressed JWE tokens (CVE-2024-28176)
 * - Enforces algorithm allowlist
 * - Enforces max clock skew of 30 seconds
 */
export async function verifyJWT(
  token: string,
  publicKey: CryptoKey | Uint8Array,
  options?: {
    audience?: string;
    issuer?: string;
  }
): Promise<JWTVerifyResult> {
  try {
    // Guard: reject tokens that look like JWE (5 dot-separated parts = encrypted)
    const parts = token.split(".");
    if (parts.length === 5) {
      // This is a JWE — check for compression header before decrypting
      // Compressed JWEs are the attack vector for CVE-2024-28176
      try {
        const header = JSON.parse(Buffer.from(parts[0], "base64url").toString("utf8"));
        if (header.zip) {
          return {
            payload: {} as RemitFlowJWTPayload,
            isValid: false,
            error: "Compressed JWE tokens are not accepted (security policy)",
          };
        }
      } catch {
        return {
          payload: {} as RemitFlowJWTPayload,
          isValid: false,
          error: "Malformed JWE header",
        };
      }
    }

    // Guard: reject oversized tokens (decompression bomb protection)
    if (token.length > MAX_JWE_PAYLOAD_BYTES) {
      return {
        payload: {} as RemitFlowJWTPayload,
        isValid: false,
        error: "Token exceeds maximum allowed size",
      };
    }

    const { payload } = await jwtVerify(token, publicKey, {
      algorithms: [...ALLOWED_ALGORITHMS],
      clockTolerance: 30, // 30 second clock skew tolerance
      maxTokenAge: MAX_TOKEN_AGE_SECONDS,
      audience: options?.audience,
      issuer: options?.issuer,
    });

    return {
      payload: payload as RemitFlowJWTPayload,
      isValid: true,
    };
  } catch (err: unknown) {
    const message = err instanceof Error ? err.message : "Unknown JWT error";
    return {
      payload: {} as RemitFlowJWTPayload,
      isValid: false,
      error: message,
    };
  }
}

/**
 * Safely decode a JWT without verification (for logging/debugging only).
 * Never use this for authorization decisions.
 */
export function decodeJWTUnsafe(token: string): JWTPayload | null {
  try {
    return decodeJwt(token);
  } catch {
    return null;
  }
}

/**
 * Extract the subject claim from a token without full verification.
 * Used only for logging — NOT for authorization.
 */
export function extractSubjectForLogging(token: string): string {
  const payload = decodeJWTUnsafe(token);
  return payload?.sub ?? "unknown";
}

File: services/python-keycloak-service/app/jwt_secure.py
"""
RemitFlow — Secure Python JWT Utilities
═══════════════════════════════════════════════════════════════════════════
Replaces python-jose with joserfc for all JWT operations.

Mitigations:
  - CVE-2024-28176: JWE Decompression DoS
    → joserfc rejects compressed JWE by default unless explicitly enabled
    → We add an explicit guard to reject any JWE with a 'zip' header
  - Algorithm confusion attacks
    → Strict algorithm allowlist enforced
  - JWT bomb attacks
    → Maximum token size enforced before parsing
"""

from __future__ import annotations

import base64
import json
import logging
from typing import Any

logger = logging.getLogger(__name__)

# Maximum token size — reject anything larger (decompression bomb protection)
MAX_TOKEN_BYTES = 65_536  # 64 KiB

# Strict algorithm allowlist — never allow "none" or symmetric algorithms
# for tokens that arrive from external sources
ALLOWED_ALGORITHMS = frozenset(["RS256", "RS384", "RS512", "ES256", "ES384", "ES512"])


def _decode_header_unsafe(token: str) -> dict[str, Any]:
    """Decode the JWT header without verification (for pre-validation only)."""
    try:
        parts = token.split(".")
        if len(parts) < 2:
            raise ValueError("Token has fewer than 2 parts")
        # Add padding if needed
        header_b64 = parts[0] + "=" * (4 - len(parts[0]) % 4)
        return json.loads(base64.urlsafe_b64decode(header_b64).decode("utf-8"))
    except Exception as exc:
        raise ValueError(f"Cannot decode token header: {exc}") from exc


def verify_token(token: str, public_key: Any, *, audience: str | None = None) -> dict[str, Any]:
    """
    Verify a JWT token using joserfc with strict security controls.

    Args:
        token: The raw JWT string
        public_key: The RSA or EC public key for verification
        audience: Expected audience claim (optional but recommended)

    Returns:
        The verified payload claims dict

    Raises:
        ValueError: If the token is invalid, expired, or uses a disallowed algorithm
        SecurityError: If the token contains a security violation (e.g., compressed JWE)
    """
    # Guard 1: Reject oversized tokens
    if len(token.encode("utf-8")) > MAX_TOKEN_BYTES:
        raise ValueError(f"Token size exceeds maximum of {MAX_TOKEN_BYTES} bytes")

    # Guard 2: Detect JWE (5 parts) and reject compressed variants (CVE-2024-28176)
    parts = token.split(".")
    if len(parts) == 5:
        try:
            header = _decode_header_unsafe(token)
            if header.get("zip"):
                raise ValueError(
                    "Security policy violation: compressed JWE tokens are not accepted "
                    "(CVE-2024-28176 mitigation)"
                )
        except ValueError:
            raise
        except Exception as exc:
            raise ValueError(f"Malformed JWE token: {exc}") from exc

    # Guard 3: Check algorithm allowlist before verification
    try:
        header = _decode_header_unsafe(token)
        alg = header.get("alg", "")
        if alg not in ALLOWED_ALGORITHMS:
            raise ValueError(
                f"Algorithm '{alg}' is not in the allowlist. "
                f"Allowed: {sorted(ALLOWED_ALGORITHMS)}"
            )
    except ValueError:
        raise

    # Perform actual verification using joserfc
    try:
        from joserfc import jwt
        from joserfc.jwk import OctKey, RSAKey, ECKey

        token_obj = jwt.decode(token, public_key)
        claims = token_obj.claims

        # Validate audience if provided
        if audience and claims.get("aud") != audience:
            raise ValueError(
                f"Token audience mismatch: expected '{audience}', got '{claims.get('aud')}'"
            )

        return claims

    except ImportError:
        # Fallback: if joserfc not yet installed, log warning and raise
        logger.error(
            "joserfc is not installed. Install with: pip install joserfc>=0.12.0"
        )
        raise RuntimeError(
            "JWT verification unavailable: joserfc package not installed"
        )
    except Exception as exc:
        logger.warning("JWT verification failed: %s", exc)
        raise ValueError(f"JWT verification failed: {exc}") from exc


def extract_subject_for_logging(token: str) -> str:
    """
    Extract the 'sub' claim from a token WITHOUT verification.
    For logging purposes only — never use for authorization.
    """
    try:
        parts = token.split(".")
        if len(parts) < 2:
            return "unknown"
        payload_b64 = parts[1] + "=" * (4 - len(parts[1]) % 4)
        payload = json.loads(base64.urlsafe_b64decode(payload_b64).decode("utf-8"))
        return str(payload.get("sub", "unknown"))
    except Exception:
        return "unknown"
